sql/src/main/java/io/crate/metadata/SystemTable.java

```diff
@@ -295,139 +295,141 @@ public void addExpression(HashMap<ColumnIdent, RowCollectExpressionFactory<T>> e
+        private final ColumnIdent baseColumn;
+        private final Function<ParentItemType, List<ItemType>> getItems;
 
-            private final P parent;
-            private final ArrayList<Column<ItemType, ?>> columns = new ArrayList<>();
-            private final ColumnIdent baseColumn;
-            private final Function<ParentItemType, List<ItemType>> getItems;
+        public ObjectArrayBuilder(P parent, ColumnIdent baseColumn, Function<ParentItemType, List<ItemType>> getItems) {
+            this.parent = parent;
+            this.baseColumn = baseColumn;
+            this.getItems = getItems;
+        }
 
-            public ObjectArrayBuilder(P parent, ColumnIdent baseColumn, Function<ParentItemType, List<ItemType>> getItems) {
-                this.parent = parent;
-                this.baseColumn = baseColumn;
-                this.getItems = getItems;
+        public P endObjectArray() {
+            ObjectType.Builder typeBuilder = ObjectType.builder();
+            ArrayList<Column<ItemType, ?>> directChildren = new ArrayList<>();
+            for (var col : columns) {
+                if (col.column.path().size() == baseColumn.path().size() + 1) {
+                    directChildren.add(col);
+                }
+            }
+            for (var column : directChildren) {
+                typeBuilder.setInnerType(column.column.leafName(), column.type);
+            }
+            ObjectType objectType = typeBuilder.build();
+            parent.add(new Column<>(baseColumn, new ArrayType<>(objectType), getLeafColumnValues(directChildren)));
+            for (var column : columns) {
+                addColumnToParent(column);
             }
+            return parent;
+        }
 
-            public P endObjectArray() {
-                ObjectType.Builder typeBuilder = ObjectType.builder();
-                ArrayList<Column<ItemType, ?>> directChildren = new ArrayList<>();
-                for (var col : columns) {
-                    if (col.column.path().size() == baseColumn.path().size() + 1) {
-                        directChildren.add(col);
+        public Function<ParentItemType, List<Map<String, Object>>> getLeafColumnValues(ArrayList<Column<ItemType, ?>> directChildren) {
+            return xs -> {
+                var items = getItems.apply(xs);
+                ArrayList<Map<String, Object>> result = new ArrayList<>(items.size());
+                for (ItemType item : items) {
+                    HashMap<String, Object> map = new HashMap<>(directChildren.size());
+                    for (int i = 0; i < directChildren.size(); i++) {
+                        Column<ItemType, ?> column = directChildren.get(i);
+                        try {
+                            Object value = column.getProperty.apply(item);
+                            map.put(column.column.leafName(), value);
+                        } catch (NullPointerException ignored) {
```

Comment:
Is there a way to avoid to catch a `NullPointerException` ?

Reply:
Yes, by making sure that all the expressions we use to retrieve the properties handle `null` cases correctly.  Not sure if that's worth it. We could follow up on it but I'd keep it for now.